econml/policy/forest/_tree.py

```diff
@@ -0,0 +1,566 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+#
+# This code contains snippets of code from:
+# https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/_classes.py
+# published under the following license and copyright:
+# BSD 3-Clause License
+#
+# Copyright (c) 2007-2020 The scikit-learn developers.
+# All rights reserved.
+
+import numpy as np
+import numbers
+from math import ceil
+from ...tree import Tree
+from ...tree._criterion import Criterion
+from ...tree._splitter import Splitter, BestSplitter
+from ...tree import DepthFirstTreeBuilder
+from ...tree import _tree
+from ..._tree_exporter import _SingleTreeExporterMixin, _PolicyTreeDOTExporter, _PolicyTreeMPLExporter
+from ._criterion import LinearPolicyCriterion
+from . import _criterion
+from sklearn.base import BaseEstimator
+from sklearn.model_selection import train_test_split
+from sklearn.utils import check_array, check_X_y
+from sklearn.utils import check_random_state
+from sklearn.utils.validation import _check_sample_weight
+from sklearn.utils.validation import check_is_fitted
+import copy
+
+# =============================================================================
+# Types and constants
+# =============================================================================
+
+DTYPE = _tree.DTYPE
+DOUBLE = _tree.DOUBLE
+
+CRITERIA_POLICY = {"neg_welfare": LinearPolicyCriterion}
+
+SPLITTERS = {"best": BestSplitter, }
+
+MAX_INT = np.iinfo(np.int32).max
+
+# =============================================================================
+# Base Policy tree
+# =============================================================================
+
+
+class PolicyTree(_SingleTreeExporterMixin, BaseEstimator):

```

Comment:
Any way we can make a common base class between this and the GRF tree? I put the two classes from `__init__` to `predict` in a diff checker and out of ~300 lines, only 60 are different and of those 60, a bunch are just docstrings. A lot of the logic, especially in fit is pretty much the same. it would make it much easier to maintain just one piece of code.

Reply:
Done!